File: docs/courses/data-structure/samples/queue/LinkedListQueue/LinkedListQueue.java
package queue.LinkedListQueue;

public class LinkedListQueue<E> {
    private Node<E> front;
    private Node<E> rear;
    private int size;

    public LinkedListQueue() {
        this.front = null;
        this.rear = null;
        this.size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public void offer(E value) {
        Node<E> node = new Node<>(value);

        if (isEmpty()) {
            front = node;
        } else {
            rear.next = node;
        }

        rear = node;
        size++;
    }

    public E poll() {
        if (isEmpty()) return null;

        E element = front.data;
        Node<E> nextNode = front.next;

        front.data = null;
        front.next = null;

        front = nextNode;
        size--;

        return element;
    }

    public E peek() {
        if (isEmpty()) return null;

        return front.data;
    }
}

class Node<E> {
    E data;
    Node<E> next;

    public Node(final E data) {
        this.data = data;
        this.next = null;
    }
}

File: docs/courses/data-structure/samples/queue/LinkedListQueue/LinkedListQueue.c
# include <stdio.h>
# include <stdlib.h>
# include <string.h>

typedef struct Node {
    int data;
    struct Node *next;
}Node;

typedef struct Queue {
    Node *front;
    Node *rear;
    int size;
}Queue;

void initQueue(Queue *queue) {
    queue -> front = queue -> rear = NULL;
    queue -> size = 0;
}

int isEmpty(Queue *queue) {
    return queue -> size == 0;
}

void offer(Queue *queue, int data) {
    Node *node = (Node *)malloc(sizeof(Node));
    node -> data = data;
    node -> next = NULL;

    if (isEmpty(queue)) {
        queue -> front = node;
    } else {
        queue -> rear -> next = node;
    }

    queue -> rear = node;
    queue -> size++;
}

int poll(Queue *queue) {
    int result = 0;
    Node *node;

    if (isEmpty(queue)) {
        return result;
    }

    node = queue -> front;
    result = node -> data;
    queue -> front = node -> next;
    free(node);
    queue -> size--;

    return result;
}

int main(void)
{
    int i;
    Queue queue;

    initQueue(&queue);                  
    for (i = 1; i <= 5; i++)            
    {
        offer(&queue, i);
    }
    while (!isEmpty(&queue))           
    {
        printf("%d ", poll(&queue));    // expected output: 1 2 3 4 5
    }
    printf("\n");
    return 0;
}